### PR TITLE
[4.0] [com_banners] Prepared statements (frontend)

### PR DESCRIPTION
**Pull request by SharkyKZ**
### Summary of Changes

This adds prepared statements and cleans up queries in `com_banners` frontend.

### Testing Instructions

Create some banners.
Publish some banner modules.
Browse around frontend.
Click on some banners.
In backend check that banner statistics (impressions and clicks) are still tracked correctly.

### Expected result

Works like before.

### Documentation Changes Required

No.

### CODE REVIEW

**Comment by 1apweb** [comment]
I have tested this item :white_check_mark: successfully on 24e7ede3ee62686e27e148eaf97229f8edc62f42<br /><br />The patch now displays the number of clicks correctly. **But** I created two banners, and when I click one, Total print displays an impression for each banner when there is only one that has been displayed.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26811">issues.joomla.org/tracker/joomla-cms/26811</a>.</sub>

**Comment by SharkyKZ** [comment]
@1apweb Sorry, I don't understand. You show 1 banner but impressions of multiple banners increase?

**Comment by 1apweb** [comment]
I've added a third banner and I realize that all three are displayed at the same time, even with the "pinned, randomize" option of the module?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26811">issues.joomla.org/tracker/joomla-cms/26811</a>.</sub>

**Comment by SharkyKZ** [comment]
That is correct. Set "Count" option to 1 if you want to display one banner at a time.

**Comment by 1apweb** [comment]
Oh sorry, the test is now ok. Thank you<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26811">issues.joomla.org/tracker/joomla-cms/26811</a>.</sub>

File: components/com_banners/Model/BannersModel.php [hunk 1]
```diff
@@ -64,83 +65,132 @@ protected function getListQuery()
...
 		{
-			$query->where('a.cid = ' . (int) $cid)
-				->where('cl.state = 1');
+			$query->where(
+				[
+					$db->quoteName('a.cid') . ' = :clientId',
+					$db->quoteName('cl.state') . ' = 1',
+				]
+			)
+				->bind(':clientId', $cid);
```
Comment (Quy) [inline components/com_banners/Model/BannersModel.php]:
```suggestion
				->bind(':clientId', $cid , ParameterType::INTEGER);
```

File: components/com_banners/Model/BannersModel.php [hunk 2]
```diff
@@ -64,83 +65,132 @@ protected function getListQuery()
...
 		{
-			$query->where('a.cid = ' . (int) $cid)
-				->where('cl.state = 1');
+			$query->where(
+				[
+					$db->quoteName('a.cid') . ' = :clientId',
+					$db->quoteName('cl.state') . ' = 1',
+				]
+			)
+				->bind(':clientId', $cid , ParameterType::INTEGER);
```
Comment (richard67) [inline components/com_banners/Model/BannersModel.php]:
```suggestion
				->bind(':clientId', $cid, ParameterType::INTEGER);
```

PHPCS

**Comment by 1apweb** [comment]
I have tested this item :white_check_mark: successfully on 7bf4db98754a1c6ed3dd4ebdedb6aea94de2ef93<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26811">issues.joomla.org/tracker/joomla-cms/26811</a>.</sub>

**Comment by 1apweb** [comment]
Nice<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26811">issues.joomla.org/tracker/joomla-cms/26811</a>.</sub>

**Comment by waader** [comment]
I have tested this item :white_check_mark: successfully on 7bf4db98754a1c6ed3dd4ebdedb6aea94de2ef93<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26811">issues.joomla.org/tracker/joomla-cms/26811</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26811">issues.joomla.org/tracker/joomla-cms/26811</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**